{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e4166360-3eff-4836-887f-e68b95b09ae9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-12-03 17:08:50--  http://d28rh4a8wq0iu5.cloudfront.net/ads1/data/phix.fa\n",
      "18.161.245.40, 18.161.245.54, 18.161.245.3, ...a8wq0iu5.cloudfront.net)... \n",
      "connected. to d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)|18.161.245.40|:80... \n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 5528 (5.4K) [application/octet-stream]\n",
      "Saving to: ‘phix.fa’\n",
      "\n",
      "phix.fa             100%[===================>]   5.40K  --.-KB/s    in 0s      \n",
      "\n",
      "2025-12-03 17:08:50 (315 MB/s) - ‘phix.fa’ saved [5528/5528]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget http://d28rh4a8wq0iu5.cloudfront.net/ads1/data/phix.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b3e0a6bb-dc91-43fa-add2-d3a972b9c45d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reading the genome\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            if not line[0] == \">\":\n",
    "                genome += line.rstrip()\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "44698745-e355-41ba-9775-81e8741442e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reading the genome file which we just downloaded\n",
    "genome = readGenome('phix.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5238054c-7dc4-4d3d-8ab7-c9bc43e44167",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Naive exact matching algorithm\n",
    "def naive(p,t):\n",
    "    occurences = []\n",
    "    for i in range(len(t) - len(p) + 1):\n",
    "        match = True\n",
    "        for j in range(len(p)):\n",
    "            if not t[i+j] == p[j]:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurences.append(i)\n",
    "    return occurences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "76cf2ad0-059b-4f79-86bb-ea44a7984146",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 5, 9]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = 'AGCTTAGATAGC'\n",
    "p = 'AG'\n",
    "naive(p,t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1cdaae9a-830a-4dc2-a681-92c0b57af07c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random \n",
    "def generateReads(genome, numReads, readLen):\n",
    "    ''' Generate reads from random positions in the given genome '''\n",
    "    reads = []\n",
    "    for _ in range(numReads):\n",
    "        start = random.randint(0, len(genome) - readLen) - 1\n",
    "        reads.append(genome[start : start + readLen])\n",
    "    return reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cff06501-088d-4707-b310-77e6763aca27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 / 100 reads matched exactly\n"
     ]
    }
   ],
   "source": [
    "#lets genrate the random reads from the genome we downloaded eariler by taking subsequences of a genome string taking from a random position\n",
    "#Example with artificial reads\n",
    "reads = generateReads(genome, 100, 100)\n",
    "\n",
    "numMatched = 0\n",
    "for r in reads:\n",
    "    matches = naive(r, genome)\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print(f\"{numMatched} / {len(reads)} reads matched exactly\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8a8277c-195e-4454-a132-994a66dad17e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets try this same example with real data.\n",
    "!wget http://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR266411_1.first1000.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cfa3f1f3-8864-4fa6-9894-503c45d22586",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 / 1000 reads matched the genome exactly\n"
     ]
    }
   ],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline() # skip name line\n",
    "            seq = fh.readline().rstrip() # read base sequence\n",
    "            fh.readline() # skip placeholder line\n",
    "            fh.readline() # skip base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "    return sequences\n",
    "\n",
    "phix_reads = readFastq('ERR266411_1.first1000.fastq')\n",
    "\n",
    "numMatched = 0\n",
    "n = 0\n",
    "for r in phix_reads:\n",
    "    matches = naive(r, genome)\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print(f\"{numMatched} / {n} reads matched the genome exactly\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c3a606ee-ba1f-4f4a-a952-0741015e81c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "459 / 1000 reads matched the genome exactly\n"
     ]
    }
   ],
   "source": [
    "''' These are sequecing reads actually from the genome and yet 7 out of 1000 actually matched!\n",
    "why?? \n",
    "1) sequencing errors\n",
    "2)differences b/w the sequences and reference genome\n",
    "3) what if the reads are coming from reverse strand of the genome\n",
    "'''\n",
    "#What if we take a part of a read an match with the genome? maybe 30 bases instead of 100\n",
    "\n",
    "numMatched = 0\n",
    "n = 0\n",
    "for r in phix_reads:\n",
    "    r = r[:30]\n",
    "    matches = naive(r, genome)\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print(f\"{numMatched} / {n} reads matched the genome exactly\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "5abd82cc-bf03-4926-b6e1-cc164735e536",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "932 / 1000 reads matched the genome exactly\n"
     ]
    }
   ],
   "source": [
    "#reverse complement of the genome\n",
    "def reverseComplement(s):\n",
    "    complement = {'A':'T', \n",
    "                  'T':'A', \n",
    "                  'G':'C',\n",
    "                  'C':'G',\n",
    "                  'N':'N'}\n",
    "    t = ''\n",
    "    for base in s:\n",
    "        t = complement[base] + t\n",
    "    return t\n",
    "    \n",
    "numMatched = 0\n",
    "n = 0\n",
    "for r in phix_reads:\n",
    "    r = r[:30]\n",
    "    matches = naive(r, genome)\n",
    "    matches.extend(naive(reverseComplement(r), genome))\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print(f\"{numMatched} / {n} reads matched the genome exactly\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e9eeba8-f146-4324-a7b6-a5cf41e97f39",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
